Always set expiry on server resource in heartbeats

awly opened the pull request:
All servers (ssh/kube/app/db) should get cleaned up some time after
deletion. If a server implements `GetServerInfo` without populating
`Expiry`, default expiry to `ServerTTL` in the heartbeat code.

If we don't clean up deleted server objects, things like kube routing
will try to reach obsolete endpoints and users will forever see deleted
kube clusters in `tsh`.

r0mant commented on lib/srv/heartbeat.go:
```diff
@@ -313,6 +313,11 @@ func (h *Heartbeat) fetch() error {
 	}
+	// Always set server expiry, no server resource should linger forever after
+	// deletion.
+	if server.Expiry().IsZero() {
```
Maybe write a unit test for this as well?

russjones commented on lib/srv/heartbeat.go:
```diff
@@ -313,6 +313,11 @@ func (h *Heartbeat) fetch() error {
 	}
+	// Always set server expiry, no server resource should linger forever after
+	// deletion.
+	if server.Expiry().IsZero() {
```
Shouldn't this be logic internal to `services.Server`? Why not add this to `CheckAndSetDefaults` for a `services.Server`?

a-palchikov replied:
I only skimmed through the implementations of `GetServerInfo` but some do return references to internal values while others generate a new value instead, so there's a chance that this will race as the heartbeat's Run is executed in a separate goroutine.
Also, it looks like only the `lib/kube/proxy#TLSServer.GetServerInfo` does not update the TTL - should it be done there instead since it has more context and can protect the state if necessary?

awly replied:
Good point @a-palchikov, didn't think about `GetServerInfo` returning data with references to shared memory.

I tried making a deep-copy on the heartbeat side, but for some reason `proto.Clone` fails.
I tried setting the expiry in `services.Server.CheckAndSetDefaults`, but it broke a bunch of tests and also resulted in logs of plumbing changes to pass `clockwork.Clock` in.

Changed the PR to only update `GetServerInfo` in `lib/kube/proxy` for now and set `Expiry` like all the other servers do. At some point, I do want to centralize this logic, since all servers use the same constant TTL.

a-palchikov replied:
> I tried making a deep-copy on the heartbeat side, but for some reason proto.Clone fails.

This might be related to [this](https://github.com/gravitational/teleport/pull/4790) work. Which error did you get from `proto.Clone`?

awly replied:
The error was also reflect-related, but different.
I don't have it on-hand right now unfortunately.

Commit:
Set TTL on kube_service resources

Without this, deleted kube_services linger in the backend and show up as
obsolete kubernetes clusters in tsh.

Ideally, this TTL logic should be enforced centrally, but I'd like to
fix the bug first, and do a larger refactoring later.
